src/TextInputWithTokens.tsx

```diff
@@ -0,0 +1,247 @@
+import React, { ComponentPropsWithoutRef, FocusEventHandler, KeyboardEventHandler, useEffect, useRef, useState } from 'react'
+import {omit} from '@styled-system/props'
+import styled from 'styled-components'
+import { FocusKeys } from './behaviors/focusZone'
+import { useCombinedRefs } from './hooks/useCombinedRefs'
+import { useFocusZone } from './hooks/useFocusZone'
+import {ComponentProps, MandateProps} from './utils/types'
+import Token, { TokenProps } from './Token/Token'
+import TokenLabel, { TokenLabelProps } from './Token/TokenLabel'
+import TokenProfile, { TokenProfileProps } from './Token/TokenProfile'
+import { TokenSizeKeys } from './Token/TokenBase'
+import TextInput, { TextInputProps } from './TextInput'
+import { useProvidedRefOrCreate } from './hooks'
+import UnstyledTextInput from './_UnstyledTextInput'
+
+const InputWrapper = styled.div`
+  order: 1;
+  flex-grow: 1;
+`
+
+type AnyTokenProps = Partial<TokenProps & TokenLabelProps & TokenProfileProps>
+type TokenDatum = MandateProps<AnyTokenProps, 'id' | 'text'>
+type TextInputWithTokensInternalProps = {
+  /**
+   * The array of tokens to render
+   */
+  tokens: TokenDatum[]
+  /**
+   * The function that gets called when a token is removed
+   */
+  onTokenRemove: (tokenId: string | number) => void
+  /**
+   * The component used to render each token
+   */
+  tokenComponent?: React.ComponentType<ComponentPropsWithoutRef<typeof Token> | ComponentPropsWithoutRef<typeof TokenLabel> | ComponentPropsWithoutRef<typeof TokenProfile>>
+  /**
+   * The maximum height of the component. If the content in the input exceeds this height,
+   * it will scroll vertically
+   */
+  maxHeight?: React.CSSProperties['maxHeight']
+  /**
+   * Whether tokens should render inline horizontally. By default, tokens wrap to new lines.
+   */
+  preventTokenWrapping?: boolean
+  /**
+   * The size of the tokens
+   */
+  tokenSizeVariant?: TokenSizeKeys
+  /**
+   * Whether the remove buttons should be rendered in the tokens
+   */
+  // TODO: confirm whether or not we're comfortable with allowing people to hide that button
+  hideTokenRemoveButtons?: boolean
+} & TextInputProps
+
+// using forwardRef is important so that other components (ex. Autocomplete) can use the ref
+const TextInputWithTokensComponent = React.forwardRef<HTMLInputElement, TextInputWithTokensInternalProps & {
+  selectedTokenIdx: number | undefined,
+  setSelectedTokenIdx: React.Dispatch<React.SetStateAction<number | undefined>>
```

Comment:
We generally prefer not to abbreviate shorter words like `index` and `event`:
```suggestion
  selectedTokenIndex: number | undefined,
  setSelectedTokenIndex: React.Dispatch<React.SetStateAction<number | undefined>>
```